File: compiler/database/span.go
package database

import (
	"fmt"
	"math"
	"slices"
	"strings"
)

type Span struct {
	Path   string   `json:"path"`
	Start  Location `json:"start"`
	End    Location `json:"end"`
	Source string   `json:"source"`
}

type Location struct {
	Line   int `json:"line"`
	Column int `json:"column"`
	Index  int `json:"index"`
}

func JoinSpans(left Span, right Span, source string) Span {
	return Span{
		Path:   left.Path,
		Start:  left.Start,
		End:    right.End,
		Source: source[left.Start.Index:max(right.End.Index, left.Start.Index)],
	}
}

func CompareSpans(left Span, right Span) int {
	if left.Path != right.Path {
		return strings.Compare(left.Path, right.Path)
	}

	if left.Start.Index != right.Start.Index {
		return left.Start.Index - right.Start.Index
	}

	if left.End.Index != right.End.Index {
		return left.End.Index - right.End.Index
	}

	return 0
}

func SpansAreEqual(left Span, right Span) bool {
	return CompareSpans(left, right) == 0
}

func HaveEqualSpans(left Node, right Node) bool {
	return SpansAreEqual(GetSpanFact(left), GetSpanFact(right))
}

func (span Span) String() string {
	return fmt.Sprintf("%s:%d:%d", span.Path, span.Start.Line, span.Start.Column)
}

func NullSpan() Span {
	return Span{
		Path:   "",
		Start:  NullLocation(),
		End:    NullLocation(),
		Source: "",
	}
}

func NullLocation() Location {
	return Location{
		Line:   1,
		Column: 1,
		Index:  0,
	}
}

func SortByProximity(nodes []Node, source Node) {
	sourceSpan := GetSpanFact(source)
	slices.SortStableFunc(nodes, func(left Node, right Node) int {
		leftSpan := GetSpanFact(left)
		rightSpan := GetSpanFact(right)

		// Deprioritize nodes located before `source`
		if leftSpan.Start.Index < sourceSpan.Start.Index {
			return math.MaxInt32
		}
		if rightSpan.Start.Index < sourceSpan.Start.Index {
			return math.MinInt32
		}

		leftDistance := sourceSpan.Start.Index - leftSpan.Start.Index
		leftDistance = max(leftDistance, -leftDistance)

		rightDistance := sourceSpan.Start.Index - rightSpan.Start.Index
		rightDistance = max(rightDistance, -rightDistance)

		return leftDistance - rightDistance
	})
}

File: compiler/src/database/span.rs
use crate::database::{Db, Fact, NodeRef, Render};
use regex::Regex;
use serde::Serialize;
use std::{
    borrow::Cow,
    fmt::{Debug, Display},
    sync::LazyLock,
};

#[derive(Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize)]
pub struct Span {
    pub path: String,
    pub start: Location,
    pub end: Location,
    pub source: String,
}

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize)]
pub struct Location {
    pub line: usize,
    pub column: usize,
    pub index: usize,
}

impl Fact for Span {}

impl Span {
    pub fn empty() -> Span {
        Span {
            path: String::new(),
            start: Location::empty(),
            end: Location::empty(),
            source: String::new(),
        }
    }

    pub fn join(left: Self, right: Self, source: &str) -> Self {
        Span {
            path: left.path.clone(),
            start: left.start.clone(),
            end: right.end.clone(),
            source: source[left.start.index..right.end.index.max(left.start.index)].to_string(),
        }
    }

    pub fn is_inside(&self, other: &Self) -> bool {
        self.path == other.path
            && self.start.index >= other.start.index
            && self.end.index <= other.end.index
    }
}

impl Display for Span {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "{}:{}:{}-{}:{}",
            self.path, self.start.line, self.start.column, self.end.line, self.end.column
        )
    }
}

impl Debug for Span {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self)
    }
}

impl Render for Span {
    fn write(&self, w: &mut dyn std::fmt::Write, _db: &Db) -> std::fmt::Result {
        write!(w, "at {}", self)
    }
}

impl Location {
    pub fn empty() -> Location {
        Location {
            line: 1,
            column: 1,
            index: 0,
        }
    }
}

impl Db {
    pub fn have_equal_spans(&self, left: &NodeRef, right: &NodeRef) -> bool {
        self.span(left) == self.span(right)
    }
}

impl Span {
    pub fn as_node_source(&self) -> String {
        static BRACES_REGEX: LazyLock<Regex> =
            LazyLock::new(|| Regex::new(r#"(?s)\{.*\}"#).unwrap());

        static ASSIGNMENT_REGEX: LazyLock<Regex> =
            LazyLock::new(|| Regex::new(r#"(?s)::?.*"#).unwrap());

        static BOUNDS_REGEX: LazyLock<Regex> =
            LazyLock::new(|| Regex::new(r#"(?s)\bwhere\b.*"#).unwrap());

        static LINES_REGEX: LazyLock<Regex> = LazyLock::new(|| Regex::new(r#"(?s)\n.*"#).unwrap());

        let source = self.as_definition_source();

        // Collapse braces
        let source = BRACES_REGEX.replace_all(&source, "{⋯}");

        let source = if let Some(left_parentheses_index) = source.find('(')
            && let Some(annotate_index) = source.find("::")
            && annotate_index < left_parentheses_index
        {
            // Remove assignments and type annotations
            ASSIGNMENT_REGEX.replace_all(&source, "")
        } else {
            source
        };

        // Remove bounds
        let source = BOUNDS_REGEX.replace_all(&source, "");

        // Collapse multiple lines
        let source = LINES_REGEX.replace_all(&source, "⋯");

        source.trim().to_string()
    }

    pub fn as_definition_source(&self) -> String {
        static ATTRIBUTES_REGEX: LazyLock<Regex> =
            LazyLock::new(|| Regex::new(r#"(\[.*\]\s*)*"#).unwrap());

        static COMMENTS_REGEX: LazyLock<Regex> = LazyLock::new(|| Regex::new(r#"--.*\n"#).unwrap());

        // Strip attributes
        let source = ATTRIBUTES_REGEX.replace_all(&self.source, "");

        // Strip comments
        let source = COMMENTS_REGEX.replace_all(&source, "");

        // Remove parentheses
        let source = if source != "()" && source.starts_with("(") && source.ends_with(")") {
            let left_count = source.chars().take_while(|&c| c == '(').count();
            let right_count = source.chars().rev().take_while(|&c| c == ')').count();

            let trim_count = left_count.min(right_count);

            Cow::Borrowed(&source[trim_count..source.len() - trim_count])
        } else {
            source
        };

        // Remove assigned value...
        let source = if let Some(index) = source.find(":") {
            // ...but not type annotations or type/trait definitions
            let rest = &source[index + 1..];

            if !rest.starts_with(":") && !rest.contains(" type ") && !rest.contains(" trait ") {
                Cow::Borrowed(&source[..index])
            } else {
                source
            }
        } else {
            source
        };

        source.trim().to_string()
    }
}
